src/userSurvey.js

```diff
@@ -190,37 +234,88 @@ export async function setSurvey(options) {
         await storageSpace.set("lastSurveyRequest", lastSurveyRequest);
         await storageSpace.set("surveyCompleted", false);
         await storageSpace.set("surveyCancelled", false);
-        await storageSpace.set("surveyId", id.generateId());
         openSurveyInNewTab();
     }
 
     // Schedule a reminder for the user
     scheduleReminderForUser();
 
-    // Set a listener for the survey completion URL
+    // Set a listener for the survey completion URL.
     browser.webRequest.onBeforeRequest.addListener(
-        () => {
-            storageSpace.set("surveyCompleted", true);
-            setPopupToNoPromptPage();
-        },
-        { urls: [ (new URL(options.surveyCompletionUrl)).href + "*" ] }
+        completeSurvey,
+        { urls: [ (new URL(surveyDetails.surveyCompletionUrl)).href + "*" ] }
     );
 
-    // Set listeners for cancel and open survey button clicks in the survey request
-    messaging.onMessage.addListener(() => {
-        storageSpace.set("surveyCancelled", true);
-        setPopupToNoPromptPage();
-    }, { type: "webScience.userSurvey.cancelSurvey" });
-    messaging.onMessage.addListener(openSurveyInNewTab, { type: "webScience.userSurvey.openSurvey" });
+    // Listeners for cancel and open survey button click only need to be added once.
+    // They do not need to be added again for subsequent calls to setSurvey.
+    if (!listenersRegistered) {
+        // Set listeners for cancel and open survey button clicks in the survey request.
+        messaging.onMessage.addListener(() => {
+            storageSpace.set("surveyCancelled", true);
+            setPopupToNoPromptPage();
+        }, { type: "webScience.userSurvey.cancelSurvey" });
+        messaging.onMessage.addListener(openSurveyInNewTab, { type: "webScience.userSurvey.openSurvey" });
+    }
+
+    listenersRegistered = true;
 }
 
 /**
  * Each study participant has a persistent survey ID, generated with
  * the id module. The ID is automatically added as a parameter to
  * the survey URL, enabling researchers to import survey data from an
- * external platform and sync it with Rally data.
+ * external platform and sync it with Rally data. This method returns the
+ * survey ID, generating it if it does not already exist.
  * @returns {string} - The participant's survey ID.
  */
 export async function getSurveyId() {
-    return await storageSpace.get("surveyId");
+    initializeStorage();
+    let surveyId = await storageSpace.get("surveyId");
+    if (surveyId === null) {
+        surveyId = id.generateId();
+        await storageSpace.set("surveyId", surveyId);
+    }
+    return surveyId;
 }
+
+/**
+ * Gets the completion status of the current survey. Can be used if a
+ * subsequent survey depends on the status of the previous survey.
+ * @returns {boolean|null} - Whether the current survey has been completed.
+ * Returns null if there is no current survey.
+ */
+export async function getSurveyCompletionStatus() {
```

Comment:
This doesn't need to be `async`, the module could cache the survey status in module-wide variables.

Reply:
I think keeping this as `async` would be beneficial because a developer could immediately make a call to this function without having to initialize the module.

Reply:
Keeping this as `async` so that no prior initialization needs to occur before calling this function.